tests/pkg_tinydtls_sock_async/dtls-server.c

```diff
@@ -0,0 +1,222 @@
+                .x = ecdsa_pub_key_x,
+                .y = ecdsa_pub_key_y,
+            },
+            .client_keys = (ecdsa_public_key_t *)other_pubkeys,
+            .client_keys_size = ARRAY_SIZE(other_pubkeys),
+        },
+    },
+};
+#else /* #ifdef CONFIG_DTLS_PSK */
+static const uint8_t psk_key_0[] = PSK_DEFAULT_KEY;
+
+static const credman_credential_t credential = {
+    .type = CREDMAN_TYPE_PSK,
+    .tag = SOCK_DTLS_SERVER_TAG,
+    .params = {
+        .psk = {
+            .key = { .s = psk_key_0, .len = sizeof(psk_key_0) - 1, },
+        },
+    },
+};
+#endif
+
+static void _close_handler(event_t *);
+
+static uint8_t _recv_buf[512];
+static sock_dtls_session_t _session;
+static sock_udp_t _udp_sock;
+static sock_dtls_t _dtls_sock;
+static tinydtls_sock_event_t _close = {
+    .super.handler = _close_handler,
+    .sock = &_dtls_sock,
+};
+static bool _active = false;
+static bool _session_used = false;
+
+static void _close_sock(sock_dtls_t *sock)
+{
+    sock_udp_t *udp_sock = sock_dtls_get_udp_sock(sock);
+
+    sock_dtls_session_destroy(sock, &_session);
+    sock_dtls_close(sock);
+    sock_udp_close(udp_sock);
+    memset(&_session, 0, sizeof(_session));
+    _active = false;
+    _session_used = false;
+    puts("Terminated server");
+}
+
+static void _close_handler(event_t *ev)
+{
+    tinydtls_sock_event_t *event = (tinydtls_sock_event_t *)ev;
+    puts("Terminating server");
+    _close_sock(event->sock);
+}
+
+static void _dtls_handler(sock_dtls_t *sock, sock_async_flags_t type, void *arg)
+{
+    sock_dtls_session_t session = { 0 };
+
+    (void)arg;
+    if (!_active) {
+        puts("Error: DTLS server is not running");
+        return;
+    }
+
+    if (type & SOCK_ASYNC_CONN_RECV) {
+        puts("Session handshake received");
+        if (sock_dtls_recv(sock, &session, _recv_buf, sizeof(_recv_buf),
+                           0) != -SOCK_DTLS_HANDSHAKE) {
+            puts("Error creating session");
+            return;
+        }
+        puts("New client connected");
+        if (!_session_used) {
+            puts("Session stored");
+            _session = session;
+            _session_used = true;
+        }
+        else {
+            puts("Unable to store session. Terminating it.");
+            sock_dtls_session_destroy(sock, &session);
+            return;
+        }
+    }
+    if (type & SOCK_ASYNC_CONN_FIN) {
+        puts("Session was destroyed by peer");
+        _session_used = false;
+    }
+    if (type & SOCK_ASYNC_CONN_RDY) {
+        puts("Session became ready");
+    }
+    if (type & SOCK_ASYNC_MSG_RECV) {
+        ssize_t res;
+
+        res = sock_dtls_recv(sock, &_session, _recv_buf, sizeof(_recv_buf), 0);
+        if (!_session_used) {
+            puts("Session is not initialized");
+            return;
+        }
+        if (res >= 0) {
+            printf("Received %d bytes -- (echo)\n", (int)res);
+            res = sock_dtls_send(sock, &_session, _recv_buf, (size_t)res, 0);
```

Comment:
If I use `session` instead of the globally set `&_session` in all instances here, I run into the error below, with `-ENOTCONN` returned. @pokgak shouldn't `sock_dtls_recv()` return the correct session for me? If not, how could you e.g. handle multiple sessions?

Reply:
yeah it should return the session. I'll look into this.

Reply:
BTW: the main reason the global variable exists is to close the session when terminated or the server is stopped. Not sure we can do anything other than app-side management for terminating on close, but maybe for `CONN_FIN` another functionality to get the terminated session correctly would make sense.

Reply:
I noticed that for the second message, two `Session handshake received` are printed. These are triggered by `CONN_RECV`. So I added some debug print to follow where `CONN_RECV` is sent from. In the first message, it is from `_event()`, while in the second message it is from `_copy_buffer()`. In `_copy_buffer()` we don't copy the new session to the output `remote` parameter of `sock_dtls_recv()` because we don't have the new session there.

Reply:
Huh? The `CONN_RECV` in `_copy_buffer()` is happening, since there are still messages in the `mbox`. It is not supposed to copy the remote. It is notifying the subscriber to the events to call `sock_dtls_recv()` to fetch the message (supposedly a `DTLS_EVENT_CONNECTED` message) and that copies the remote. The call here is in response to a data receive (`MSG_RECV`), so the session _should_ be copied after the `sock_udp_recv()` somewhere.

Reply:
> If I use `session` instead of the globally set `&_session` in all instances here, I run into the error below, with `-ENOTCONN` returned. @pokgak shouldn't `sock_dtls_recv()` return the correct session for me? If not, how could you e.g. handle multiple sessions?

Can you check if this still occurs with the suggestion in https://github.com/RIOT-OS/RIOT/pull/12907#discussion_r441678942?